### PR TITLE
Support Custom Timeout for Queries

### LINKED ISSUE #1399: Custom timeouts per query
As per IRC with @juliusv 

Initial context was my desire to run `count({job=~".+"}) / count(up{job=~".+"}) by (job)` which ran into a timeout for obvious reasons.

### PR DESCRIPTION
**Pull request by Gouthamve**
This is a WIP, tests are yet to be added. Just wanted to get feedback if this was the right approach.

Some tests I added were failing as a lot of time (>30secs) was being spent here: https://github.com/prometheus/prometheus/blob/master/storage/local/storage.go#L547 and context was ignored there. Any reason for ignoring context?

Closes #1399

### CODE REVIEW

**Comment by fabxc** [comment]
Generally contexts are per request. The timeout the query engine wraps the provided context with is merely a upper bound applied to guard against too-high user-defined timeouts. Whether this should be set in the query engine or rather the client boundary (e.g. API or rule evaluation code) is a different question :)

For the introduction of custom timeouts though, this should not happen in the query engine itself but rather by the caller setting a timeout on the context passed to `Query.Exec(ctx)`.
That would make things a bit simpler too, as we'd not have to change anything in the usage of templating and rule evaluation.

**Comment by Gouthamve** [comment]
```
// WithTimeout returns a copy of parent whose Done channel is closed as soon as
// parent.Done is closed, cancel is called, or timeout elapses. The new
// Context's Deadline is the sooner of now+timeout and the parent's deadline, if
// any. If the timer is still running, the cancel function releases its
// resources.
```
Did not know that before! Will make a lot of things easy!

**Comment by Gouthamve** [comment]
@fabxc Fixed it.

But, I find sometimes that I am stuck here: https://github.com/prometheus/prometheus/blob/master/storage/local/storage.go#L547 for more than 10mins when running a heavy server. Just trying to understand why we are not cancelling and cleaning up midway.

**Comment by fabxc** [comment]
Yes, that code path should respect cancelation as well. Just nobody got around to making use of the cancelation signal throughout that bit of code yet.

**Comment by fabxc** [comment]
Code looks okay in general. @brian-brazil final concerns on HTTP API extension?

File: web/api/v1/api.go [hunk 1]
```diff
@@ -345,6 +369,7 @@ func (api *API) dropSeries(r *http.Request) (interface{}, *apiError) {
 	return res, nil
 }
 
+// Target has the information for 1 target.
```
Comment (fabxc) [inline web/api/v1/api.go]: 1 -> one

File: web/api/v1/api.go [hunk 2]
```diff
@@ -386,10 +412,12 @@ func (api *API) targets(r *http.Request) (interface{}, *apiError) {
 	return res, nil
 }
 
+// AlertmanagerDiscovery has all the active alert-managers
```
Comment (fabxc) [inline web/api/v1/api.go]: "Alertmanagers" is how we write it.
Sentence should end with period.

File: web/api/v1/api.go [hunk 3]
```diff
@@ -386,10 +412,12 @@ func (api *API) targets(r *http.Request) (interface{}, *apiError) {
...
 type AlertmanagerDiscovery struct {
 	ActiveAlertmanagers []*AlertmanagerTarget `json:"activeAlertmanagers"`
 }
 
+// AlertmanagerTarget has info on 1 AM
```
Comment (fabxc) [inline web/api/v1/api.go]: 1 -> one, terminate with period.

**Comment by brian-brazil** [comment]
Seems fine from an API standpoint.

I've been considering having scrapes pass down their timeout via a header, and remote read could do the same, but I think those APIs will necessarily be different.

**Comment by fabxc** [comment]
Okay, this one seems good to go then. 👍

**Pull request merged**